datalad/distribution/tests/test_install.py

```diff
@@ -674,6 +674,11 @@ def test_install_skip_list_arguments(src, path, path_outside):
     assert_result_count(
         result, 1, status='impossible', message="path does not exist",
         path=opj(ds.path, 'not_existing'))
+    # TODO unclear to me why this error should be like this
+    # we are asking install to pull from multiple sources, this one doesn't
+    # work, but not because it is not "associated" with a dataset, but
+    # because it doesn't exist. If it would, we would expect it to
+    # install the source as a subdataset, wouldn't we?
     assert_result_count(
         result, 1, status='error',
         message=("path not associated with dataset %s", ds),
```

Comment:
Do you agree with this assessment?

Reply:
I don't think so (or I misread that comment).
We are not calling `install -d . path_outside`, so we could conclude that `path_outside` is in fact a source. Rather we say explicitly that this the `path` to install. If it is not associated with a dataset and we didn't give a source, then we cannot deal with that, because it's not in a dataset (where we then could conclude a source from).

Reply:
Then again - I'm not quite clear what you mean by "we are asking install to pull from multiple sources".
How so? The install call doesn't say anything about sources. It's pretty much a get for a list of things, isn't it?

Reply:
> We are not calling `install -d . path_outside`, so we could conclude that `path_outside` is in fact a source.

Yes, we do `ds.install(...)`

> Rather we say explicitly that this the `path` to install. If it is not associated with a dataset and we didn't give a source, then we cannot deal with that, because it's not in a dataset (where we then could conclude a source from).

That is not have `install` works. It aims to know what you want. If you check the original issue, such multisource calls are explicitly desired. And unless `--source` is given, any path arg could be a source.

Reply:
IOW how is this different from:

```
datalad install -d . ///openfmri/ds00000{1,2} existing_subds
```

?

Reply:
Man, that's screwed even more than I remember. ;-)

But still. That would mean, that if it existed we would want to treat it as a source and install from there. However, if that path doesn't point to *some* dataset, what would you install then from there? It would still fail, because that thing is nothing we can install, since it's not associated with a dataset. Or what else do I miss?

Reply:
Yes, it should fail, just not with that error. Instead it should say

```
{
  "action": "install",
  "path": "/tmp/datalad_temp_test_install_skip_list_argumentscn7klenf/datalad_temp_test_install_skip_list_argumentsc8_dxvo2",
  "type": "dataset",
  "refds": "/tmp/datalad_temp_test_install_skip_list_argumentscn7klenf",
  "status": "error",
  "message": [
   "Failed to clone from all attempted sources: %s",
   [
    "/tmp/datalad_temp_test_install_skip_list_argumentsc8_dxvo2"
   ]
  ],
  "source_url": "/tmp/datalad_temp_test_install_skip_list_argumentsc8_dxvo2"
 },
```

which is what happens now with this PR.

It is just that I cannot fully figure out what is desired behavior, hence I appreciate the feedback.

Reply:
I see. I read the comment as "it should fail with something like `path doesn't exist`".
That new error is better - I agree.

Reply:
"Failed to clone from all attempted sources:"
nope.  
Current behavior (for better or for worse) is that whenever install gets a path as an argument, it is not taken as a source, but as a destination. Again the hybrid nature of datalad as "distribution"+VCS, so we can do e.g. `datalad install subds*` or `datalad ... search | xargs datalad install` etc.
 
If destination is already installed -- all good, IIRC we just would verify (where applicable) if it is the right one (source/version), etc.
Here because there  `-d` (some time in the past at 6957e9ab4835422d361356851ad4ee109acb524a or before) it was made forbidden to install paths outside of the explicitly specified dataset, and hence the message.  It is reall not that it cannot clone **from** that location (regardless either it exists or not).

Reply:
Oh wow. That is a revelation for me. And indeed:

```
(datalad3-dev) 127 mih@meiner /tmp % datalad create 11
[INFO   ] Creating a new annex repo at /tmp/11 
create(ok): /tmp/11 (dataset)
(datalad3-dev) mih@meiner /tmp % cd 11
(datalad3-dev) mih@meiner /tmp/11 (git)-[master] % datalad install -d . ~/forrest/collection/phase2
[WARNING] path not part of the reference dataset at /tmp/11 [get(/home/mih/forrest/collection/phase2)] 
get(impossible): ../../home/mih/forrest/collection/phase2 (dataset) [path not part of the reference dataset at /tmp/11]
(datalad3-dev) 1 mih@meiner /tmp/11 (git)-[master] % datalad --version
datalad 0.11.8
```

I am now at a complete loss of how this shall work. If a path given to `install` is always a destination, it can only ever be a location in an existing dataset. So it is really just a `get` that figures out the dataset to operate on.

Reply:
FTR: This means that `datalad -f '{path}' search -d /// haxby | xargs datalad install` (which I found a useful pattern to get matching datasets for a desired analysis is actually not applicable for a YODA scenario. I thought it would install all matching dataset in PWD, such that I could `save` them in a YODA dataset as inputs (or use `install -d .` right away). But it doesn't do that. I installs them scattered in their respective hierarchy locations under `~/datalad...`. Which is not helpful for setting up a YODA dataset. Going one by one via `install -s ///something dest` is the only working approach, which `clone` does as well.

Reply:
Shouldn't that be `datalad -f '{path}' search -d /// haxby | xargs datalad install -s` then? Emphasize on the `-s`.

Reply:
those are not sources -- those are destinations, so they should not be listed in `-s` unless you would like to take those found (but they might not even be present, so would not even work then) and install in another location.  I never wanted/needed to do that yet, and it would have not worked since `datalad install -s SRC [DEST]` can take only a single SRC at the moment.

Reply:
> unless you would like to take those found (but they might not even be present
Yes. I think that is what Michael had in mind and referred to as the YODA scenario. While that wouldn't work ATM, this could be the way to make it work (allow a list of sources as well). That way, with `-s` it would do, what Michael wants and without it would do, what you apparently had in mind when pointing out that usecase originally.

Reply:
I see.
So we have a new scenario which we have not originally envisioned or even encountered before.  I think that adding a new option `--sources` with `nargs="+"` to do `install DESTPREFIX --sources PATH1 [PATH2]` to complement `--source SRC [DEST]` would be the least ground shaking way to achieve it without breaking any compatibility.  Alternatively `--source` could be "broken" (`--source SRC DEST` would no longer work as desired; would require reordering but even then work differently since DEST would not be DESTPREFIX) to become `nargs="+"` and then `[DEST]` becoming the `[DESTPREFIX]`.  

So, I guess it was not yet even a single use case so far requiring such "multiple sources" functionality since we detected it only this late in the game, not sure if worth breaking API just to fulfill it or breeding new commands (as suggested in #3922).  
Related: we might also like to re-consider https://github.com/datalad/datalad/issues/3233 then (which would not even require any breakage ;-) )